token-group/example/src/processor.rs

```diff
@@ -0,0 +1,225 @@
+//! Program state processor
+
+use {
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        msg,
+        program_error::ProgramError,
+        program_option::COption,
+        pubkey::Pubkey,
+    },
+    spl_pod::optional_keys::OptionalNonZeroPubkey,
+    spl_token_2022::{extension::StateWithExtensions, state::Mint},
+    spl_token_group_interface::{
+        error::TokenGroupError,
+        instruction::{
+            InitializeGroup, TokenGroupInterfaceInstruction, UpdateGroupAuthority,
+            UpdateGroupMaxSize,
+        },
+        state::{Group, Member},
+    },
+    spl_type_length_value::state::TlvStateMut,
+};
+
+fn check_update_authority(
+    update_authority_info: &AccountInfo,
+    expected_update_authority: &OptionalNonZeroPubkey,
+) -> Result<(), ProgramError> {
+    if !update_authority_info.is_signer {
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+    let update_authority = Option::<Pubkey>::from(*expected_update_authority)
+        .ok_or(TokenGroupError::ImmutableGroup)?;
+    if update_authority != *update_authority_info.key {
+        return Err(TokenGroupError::IncorrectAuthority.into());
+    }
+    Ok(())
+}
+
+/// Processes a [InitializeGroup](enum.GroupInterfaceInstruction.html)
+/// instruction for a `Collection`
+pub fn process_initialize_collection(
```

Comment:
Yes, see my comment above. I intentionally use collection here to indicate that I’m using the "group" component of the group interface as a collection. 

If I introduce a program that does other types of groups, I can distinguish them even though they use the same instructions/state (see the Editions example in the stack)

Reply:
Ohhhh I see. To be honest, I think that will confuse users of the interface. Here, a program implements `TokenGroupInterface::InitializeGroup`, but when someone sends an instruction to the program they'll see `InitializeCollection` in the logs.

If I were working on a downstream program or website that calls into the interface, I would be very confused and think that either I did the wrong thing, or the implementing program did the wrong thing. To take it further, it would be like creating a `Transfer` interface, and then calling it `Reassign` in the program.

These concepts will be confusing enough, so let's encourage people to be as clear as possible when implementing other interfaces. At least, that's where my head is at. What do you think?

Reply:
I think I'm on the other side of this one. If I was a downstream app calling into Joe's Collection Program, I would be more confused if I saw `InitializeGroup` than `InitializeCollection` (although admittedly not _that_ confused).

Furthermore, if I was calling into Joe's NFT Collections and Editions Program, I'd want to make sure I'm calling the right instruction to `InitializeCollection` or `InitializeOriginalEdition`, rather than getting `InitializeGroup` for both.

Maybe I'm going beyond the scope of what this example aims to serve, but the idea (in my mind) was to demonstrate how this general interface can serve multiple use cases - including those we're familiar with today. Somebody should be able to write a program that implements `TokenMetadata` and `TokenGroup`, where `TokenGroup` is implemented more than once to accomplish more than one kind of grouping.

However, for this example specifically, I'm willing to dial it back a bit.

Reply:
In my mind, for the example program that accompanies an interface, we should do the simplest possible implementation for people to fork. If there's other inconsistent language, people will be confused.

If we want to show how to implement multiple interfaces, let's have a separate example that very clearly outlines that goal. Otherwise, it just looks like someone changed names halfway through implementation and forgot to change it everywhere, to me at least.

We should be careful to separate people who intend to use an interface (e.g. just this example), and people who intend to use a specific implementation (e.g. Joe's multi-interface program). For the example program, let's focus on the interface.

Reply:
All right that's fair. I'm going to re-word the code and also this PR.